Bigtable/tests/Unit/SmartRetriesTest.php

```diff
@@ -0,0 +1,735 @@
+            );
+        $mutations = $this->generateMutations(1, 5);
+        $entries = $this->generateEntries(1, 5);
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $entries, ['retries' => 5] + $this->options)
+            ->shouldBeCalledTimes(6)
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        $this->table->mutateRows($mutations, ['retries' => 5]);
+    }
+
+    public function testMutateRowsOnlyRetriesOnlyFailed()
+    {
+        $this->serverStream->readAll()
+            ->shouldBeCalledTimes(2)
+            ->willReturn(
+                $this->arrayAsGeneratorWithException(
+                    $this->getMutateRowsResponse(5, [2 => Code::ABORTED, 3 => Code::ABORTED])
+                ),
+                $this->arrayAsGeneratorWithException(
+                    $this->getMutateRowsResponse(1)
+                )
+            );
+        $entries = $this->generateEntries(0, 5);
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $entries, $this->options)
+            ->shouldBeCalledTimes(1)
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        $entries = $this->generateEntries(2, 3);
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $entries, $this->options)
+            ->shouldBeCalled()
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        $mutations = $this->generateMutations(0, 5);
+        $this->table->mutateRows($mutations);
+    }
+
+    public function testMutateRowsExceptionShouldPendingMutations()
+    {
+        $this->serverStream->readAll()
+            ->shouldBeCalledTimes(2)
+            ->willReturn(
+                $this->arrayAsGeneratorWithException(
+                    $this->getMutateRowsResponse(3, [1 => Code::ABORTED, 2 => Code::ABORTED])
+                ),
+                $this->arrayAsGeneratorWithException(
+                    [],
+                    $this->nonRetryingApiException
+                )
+            );
+        $entries = $this->generateEntries(0, 5);
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $entries, $this->options)
+            ->shouldBeCalledTimes(1)
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        $entries = array_merge($this->generateEntries(1, 2), $this->generateEntries(4, 5));
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $entries, $this->options)
+            ->shouldBeCalled()
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        try {
+            $mutations = $this->generateMutations(0, 5);
+            $this->table->mutateRows($mutations);
+            $this->fail('Expected exception is not thrown');
+        } catch (BigtableDataOperationException $ex) {
+            $expectedFailedMutations = [];
+            foreach (array_merge(range(1, 2), range(4, 5)) as $rowKey) {
```

Comment:
I think it the following would make the intention a bit clearer:

```suggestion
            foreach ([1, 2, 4, 5] as $rowKey) {
```